README.md
newspaper-delayed-batch-alerter
===============================

When data from a batch roundtrip is receieved, it is expected that it will be approved within a fixed period of time, for
example 20 days. This application is an autonomous component which sends a warning email if a longer period elapses
between data being received and the batch being approved.

## Configuration

In addition to the standard configuration parameters for autonomous components, this component requires the following
properties:

    #The number of days after receiving data at which to send an email alert
    delay.alert.days=

    #A comma-separated list of addresses to which to send emails
    delay.alert.email.addresses=

    #The smtp server to use for sending emails
    delay.alert.smtp.host=

    #The smtp port of the host
    delay.alert.smtp.port=


    #The "from" address to use in sending emails
    delay.alert.email.from.address=

In addition, the event-properties must have the following values:

    autonomous.pastSuccessfulEvents=Data_Received
    autonomous.oldEvents=
    autonomous.futureEvents=Roundtrip_Approved,Warning_Email_Sent




CHANGELOG.md
1.6
* Rename start script to pollAndWork.sh to match all other components

1.5
* Do not depend on an unused iterator
* Use newest version of item event framework. No functional changes for this module.
* Configuration has been extended and changed and example config has been updated. Please update your configuration files.

1.4
* Update to version 1.10 of batch event framework
* Remove dependency on mfpak integration

1.3
* Update to version 1.8 of batch event framework to fix multithreading issue

1.2
* Use MFPAK-trigger
* Use newest version of batchevent framework, which supports other trigger mechanisms
* Code cleanup
* Updated to newspaper-parent 1.2

1.1
* Fix logname

## 1.0
Initial release


src/test/java/dk/statsbiblioteket/newspaper/delayalerter/DelayAlerterComponentTest.java
package dk.statsbiblioteket.newspaper.delayalerter;

import com.icegreen.greenmail.util.GreenMail;
import com.icegreen.greenmail.util.GreenMailUtil;
import com.icegreen.greenmail.util.ServerSetup;
import dk.statsbiblioteket.medieplatform.autonomous.Batch;
import dk.statsbiblioteket.medieplatform.autonomous.Event;
import dk.statsbiblioteket.medieplatform.autonomous.ResultCollector;
import org.testng.annotations.AfterTest;
import org.testng.annotations.BeforeTest;
import org.testng.annotations.Test;

import javax.mail.MessagingException;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Properties;

import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyList;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.*;
import static org.testng.Assert.assertFalse;
import static org.testng.Assert.assertTrue;
import static org.testng.Assert.fail;

/**
 *
 */
public class DelayAlerterComponentTest {

    GreenMail greenMail;

    /**
     * Test that no mail is sent if the roundtrip is less than the maximum allowed age.
     * @throws Exception
     */
    @Test
    public void testdoWorkOnItemNotYetDelayed() throws Exception {
        Properties properties = new Properties();
        properties.setProperty(DelayAlerterConfigConstants.DELAY_ALERT_EMAIL_ADDRESSES, "foo@bar.com, bar@bar.com");
        properties.setProperty(DelayAlerterConfigConstants.DELAY_ALERT_DAYS, "20");
        final SimpleMailer simpleMailer = mock(SimpleMailer.class);
        DelayAlerterComponent component = new DelayAlerterComponent(properties, simpleMailer);
        Batch batch = new Batch();
        ResultCollector resultCollector = new ResultCollector("foo", "bar");
        Event event = new Event();
        Date now = new Date();
        event.setDate(new Date(now.getTime() - 10*24*3600*1000L));  //ten days ago
        event.setEventID("Data_Received");
        List<Event> events = new ArrayList<>();
        events.add(event);
        batch.setEventList(events);
        batch.setBatchID("B403485748392");
        batch.setRoundTripNumber(4);
        component.doWorkOnItem(batch, resultCollector);
        verify(simpleMailer, never()).sendMail(anyList(), anyString(), anyString());
    }

    /**
     * Test that a mail is sent if the roundtrip is older than the maximum allowed.
     * @throws Exception
     */
    @Test
    public void testdoWorkOnItemDelayed() throws Exception {
        Properties properties = new Properties();
        properties.setProperty(DelayAlerterConfigConstants.DELAY_ALERT_EMAIL_ADDRESSES, "foo@bar.com, bar@bar.com");
        properties.setProperty(DelayAlerterConfigConstants.DELAY_ALERT_DAYS, "20");
        final SimpleMailer simpleMailer = mock(SimpleMailer.class);
        DelayAlerterComponent component = new DelayAlerterComponent(properties, simpleMailer);
        Batch batch = new Batch();
        ResultCollector resultCollector = new ResultCollector("foo", "bar");
        Event event = new Event();
        Date now = new Date();
        event.setDate(new Date(now.getTime() - 30*24*3600*1000L));  //30 days ago
        event.setEventID("Data_Received");
        List<Event> events = new ArrayList<>();
        events.add(event);
        batch.setEventList(events);
        batch.setBatchID("B403485748392");
        batch.setRoundTripNumber(4);
        component.doWorkOnItem(batch, resultCollector);
        verify(simpleMailer, times(1)).sendMail(anyList(), anyString(), anyString());
    }

    @BeforeTest
    public void setUp() {
        ServerSetup serverSetup = new ServerSetup(40025, ServerSetup.SMTP.getBindAddress(), ServerSetup.SMTP.getProtocol());
        this.greenMail = new GreenMail(serverSetup);
        greenMail.stop();
        greenMail.start();
    }

    @AfterTest
    public void tearDown() {
        greenMail.stop();
    }


    /**
     * Test that a mail is sent if the roundtrip is older than the maximum allowed. Uses GreenMail so
     * mail is actually "sent".
     * @throws Exception
     */
    @Test
    public void testdoWorkOnItemDelayedRealMail() throws Exception {
        Properties properties = new Properties();
        properties.setProperty(DelayAlerterConfigConstants.DELAY_ALERT_EMAIL_ADDRESSES, "foo@bar.com, bar@bar.com");
        properties.setProperty(DelayAlerterConfigConstants.DELAY_ALERT_DAYS, "20");
        final SimpleMailer simpleMailer = new SimpleMailer("me@test.com", "localhost", "40025");
        DelayAlerterComponent component = new DelayAlerterComponent(properties, simpleMailer);
        Batch batch = new Batch();
        ResultCollector resultCollector = new ResultCollector("foo", "bar");
        Event event = new Event();
        Date now = new Date();
        event.setDate(new Date(now.getTime() - 30*24*3600*1000L));  //30 days ago
        event.setEventID("Data_Received");
        List<Event> events = new ArrayList<>();
        events.add(event);
        batch.setEventList(events);
        batch.setBatchID("B403485748392");
        batch.setRoundTripNumber(4);
        component.doWorkOnItem(batch, resultCollector);
        String body = GreenMailUtil.getBody(greenMail.getReceivedMessages()[0]);
        assertTrue(body.contains(batch.getFullID()));
    }

    /**
     * Test that if the mail cannot be sent then the result is not preserved. This is important as otherwise
     * there will be no future mails either. But how will we know that it failed?
     * @throws Exception
     */
    @Test
    public void testdoWorkOnItemMailFailed() throws Exception {
        Properties properties = new Properties();
        properties.setProperty(DelayAlerterConfigConstants.DELAY_ALERT_EMAIL_ADDRESSES, "foo@bar.com, bar@bar.com");
        properties.setProperty(DelayAlerterConfigConstants.DELAY_ALERT_DAYS, "20");
        final SimpleMailer simpleMailer = mock(SimpleMailer.class);
        DelayAlerterComponent component = new DelayAlerterComponent(properties, simpleMailer);
        Batch batch = new Batch();
        ResultCollector resultCollector = new ResultCollector("foo", "bar");
        Event event = new Event();
        Date now = new Date();
        event.setDate(new Date(now.getTime() - 30*24*3600*1000L));  //30 days ago
        event.setEventID("Data_Received");
        List<Event> events = new ArrayList<>();
        events.add(event);
        batch.setEventList(events);
        batch.setBatchID("B403485748392");
        batch.setRoundTripNumber(4);
        doThrow(new MessagingException("foobar")).when(simpleMailer).sendMail(anyList(), anyString(), anyString());
        try {
            component.doWorkOnItem(batch, resultCollector);
            fail("Should have thrown an exception here.");
        } catch (Exception e) {
            //expected
        }
        verify(simpleMailer, times(1)).sendMail(anyList(), anyString(), anyString());
        assertFalse(resultCollector.isPreservable());
    }



}


src/test/java/dk/statsbiblioteket/newspaper/delayalerter/DelayAlerterComponentTestIT.java
package dk.statsbiblioteket.newspaper.delayalerter;

import com.icegreen.greenmail.util.GreenMail;
import com.icegreen.greenmail.util.GreenMailUtil;
import com.icegreen.greenmail.util.ServerSetup;

import dk.statsbiblioteket.medieplatform.autonomous.*;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import javax.mail.internet.MimeMessage;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.util.*;

import static org.testng.Assert.*;

/**
 *
 */
public class DelayAlerterComponentTestIT {

    public static final String IT_EVENT = "IT_Event";
    public static final String ROUNDTRIP_APPROVED = "Roundtrip_Approved";
    protected static final String WARNING_EMAIL_SENT = "Warning_Email_Sent";
    protected static final String MANUALLY_STOPPED = "Manually_stopped";
    private final String DATA_RECEIVED = "Data_Received";


    public static Logger logger = LoggerFactory.getLogger(DelayAlerterComponentTestIT.class);

    long sleep = 10000L;
    int nsleeps = 0;
    int maxSleeps = 100;

    private NewspaperDomsEventStorage domsEventClient;
    private String batchId = "321123";
    private int roundTrip = 12;
    private EventTrigger<Batch> sboi;
    private String pathToProperties;
    private Properties properties;
    private GreenMail greenMail;
    private String pid;

    @BeforeMethod(groups = "integrationTest")
    public void setUp() throws Exception {
        logger.debug("Doing setUp.");
        String genericProperties = System.getProperty("integration.test.newspaper.properties");
        File specificProperties = new File(new File(genericProperties).getParentFile(), "newspaper-delayed-batch-alerter-config/integration.test.newspaper.properties");
        logger.debug("Loading properties from " + specificProperties.getAbsolutePath());
        properties = new Properties();
        properties.load(new FileInputStream(specificProperties));
        pathToProperties = specificProperties.getAbsolutePath();
        NewspaperDomsEventStorageFactory domsEventClientFactory = new NewspaperDomsEventStorageFactory();
        domsEventClientFactory.setFedoraLocation(properties.getProperty(ConfigConstants.DOMS_URL));
        domsEventClientFactory.setUsername(properties.getProperty(ConfigConstants.DOMS_USERNAME));
        domsEventClientFactory.setPassword(properties.getProperty(ConfigConstants.DOMS_PASSWORD));
        domsEventClientFactory.setPidGeneratorLocation(properties.getProperty(ConfigConstants.DOMS_PIDGENERATOR_URL));
        logger.debug("Creating doms client");
        domsEventClient = domsEventClientFactory.createDomsEventStorage();
        String summaLocation = properties.getProperty(ConfigConstants.AUTONOMOUS_SBOI_URL);
        PremisManipulatorFactory<Batch> factory = new PremisManipulatorFactory<>( PremisManipulatorFactory.TYPE,new BatchItemFactory());
        logger.debug("Creating sboi client");
        sboi = new NewspaperSBOIEventStorage(summaLocation, factory, domsEventClient, Integer.parseInt(properties.getProperty(ConfigConstants.SBOI_PAGESIZE,"100")));
        logger.debug("Creating round-trip object (if necessary).");
        pid = domsEventClient.createBatchRoundTrip(Batch.formatFullID(batchId, roundTrip));
        logger.debug("Created doms object {}.", pid);
        logger.debug("Resetting doms round-trip object state");
        domsEventClient.triggerWorkflowRestartFromFirstFailure(new Batch(batchId, roundTrip), "Data_Received");
        logger.debug("Waiting for reindexing.");
        waitForEvent(batchId, roundTrip, DATA_RECEIVED, false);
        ServerSetup serverSetup = new ServerSetup(40026, ServerSetup.SMTP.getBindAddress(), ServerSetup.SMTP.getProtocol());
        this.greenMail = new GreenMail(serverSetup);
        greenMail.stop();
        greenMail.start();
    }



    @AfterMethod(groups = "integrationTest")
    public void tearDown() throws Exception {
        logger.debug("Doing tearDown.");
        greenMail.stop();
    }


    /**
     * Test that we send an email on a delayed batch.
     * @throws IOException
     * @throws CommunicationException
     */
    @Test(groups = "integrationTest")
    public void testDoMainSendAlert() throws Exception {
        logger.debug("Entering testDoMainSendAlert");
        Date now = new Date();
        Date thirtyDaysAgo = new Date(now.getTime() - 30 * 24 * 3600 * 1000L);
        domsEventClient.appendEventToItem(new Batch(batchId, roundTrip), "me", thirtyDaysAgo, "details", DATA_RECEIVED, true);
        domsEventClient.appendEventToItem(new Batch(batchId, roundTrip), "me", now, "details", IT_EVENT, true);
        logger.debug("Waiting for batch to be added to SBOI");
        waitForBatchIsInSboi(batchId,
                roundTrip,
                DATA_RECEIVED + "," + IT_EVENT, ROUNDTRIP_APPROVED + "," + WARNING_EMAIL_SENT + "," + MANUALLY_STOPPED);
        DelayAlerterComponent.doMain(new String[]{"-c", pathToProperties});
        MimeMessage[] receivedMessages = greenMail.getReceivedMessages();
        //There could be other batches that trigger emails so check that there is one from us
        boolean found = false;
        for (MimeMessage message: receivedMessages) {
            if (GreenMailUtil.getBody(message).contains(batchId)) {
                found = true;
            }
        }
        assertTrue(found);
        assertTrue(batchContainsEvent(batchId, roundTrip, DelayAlerterComponent.EMAIL_SENT_EVENT));
    }

    private void waitForBatchIsInSboi(String batchId, int roundTrip, String pastEvents, String futureEvents) throws InterruptedException {
        Batch localBatch = new Batch(batchId, roundTrip);
        nsleeps = 0;
        while (true) {
            boolean eventPresent = false;
            Iterator<Batch> batchIterator;
            try {
                EventTrigger.Query<Batch> query = new EventTrigger.Query<>();
                query.getPastSuccessfulEvents().addAll(Arrays.asList(pastEvents.trim().split(",")));
                query.getFutureEvents().addAll(Arrays.asList(futureEvents.trim().split(",")));
                query.getItems().add(localBatch);
                batchIterator = sboi.getTriggeredItems(query);
                while (batchIterator.hasNext()) {
                    final Batch batch = batchIterator.next();
                    if (batch.getBatchID().equals(batchId) && batch.getRoundTripNumber().equals(roundTrip)) {
                        eventPresent = true;
                    }
                }

                if (eventPresent) {
                    return;
                }
            } catch (CommunicationException e) {
                logger.debug(e.getMessage()); //Expected during reindexing.
            }
            Thread.sleep(sleep);
            nsleeps++;
            logger.debug("nsleeps = {}/{}.", nsleeps, maxSleeps);
            if (nsleeps > maxSleeps)
                throw new RuntimeException("SBOI not updated after " + maxSleeps * sleep / 1000L + " seconds");
        }
    }

    /**
     * Test that we don't send an email on an approved batch.
     * @throws IOException
     * @throws CommunicationException
     */
    @Test(groups = "integrationTest")
    public void testDoMainApproved() throws Exception {
        logger.debug("Entering testDoMainApproved.");
        Date now = new Date();
        Date thirtyDaysAgo = new Date(now.getTime() - 30 * 24 * 3600 * 1000L);
        domsEventClient.appendEventToItem(new Batch(batchId, roundTrip), "me", thirtyDaysAgo, "details", DATA_RECEIVED, true);
        domsEventClient.appendEventToItem(new Batch(batchId, roundTrip), "me", now, "details", IT_EVENT, true);
        domsEventClient.appendEventToItem(new Batch(batchId, roundTrip), "me", new Date(), "details", ROUNDTRIP_APPROVED, true);
        waitForBatchIsInSboi(batchId,
                roundTrip,
                DATA_RECEIVED + "," + IT_EVENT + "," + ROUNDTRIP_APPROVED, WARNING_EMAIL_SENT + "," + MANUALLY_STOPPED);
        DelayAlerterComponent.doMain(new String[]{"-c", pathToProperties});
        MimeMessage[] receivedMessages = greenMail.getReceivedMessages();
        //There could be other batches that trigger emails so check that there isn't one from us
        boolean found = false;
        for (MimeMessage message: receivedMessages) {
            if (GreenMailUtil.getBody(message).contains(batchId)) {
                found = true;
            }
        }
        assertFalse(found);
        assertFalse(batchContainsEvent(batchId, roundTrip, DelayAlerterComponent.EMAIL_SENT_EVENT));
    }

    /**
     * Test that we don't send an email when a batch isn't old enough to be delayed.
     * @throws IOException
     * @throws CommunicationException
     */
    @Test(groups = "integrationTest")
    public void testDoMainNotTooOld() throws Exception {
        logger.debug("Entering testDoMainNotTooOld.");
        Date now = new Date();
        Date tenDaysAgo = new Date(now.getTime() - 10 * 24 * 3600 * 1000L);
        domsEventClient.appendEventToItem(new Batch(batchId, roundTrip), "me", tenDaysAgo, "details", DATA_RECEIVED, true);
        domsEventClient.appendEventToItem(new Batch(batchId, roundTrip), "me", now, "details", IT_EVENT, true);
        waitForBatchIsInSboi(batchId,
                roundTrip,
                DATA_RECEIVED + "," + IT_EVENT, ROUNDTRIP_APPROVED+","+WARNING_EMAIL_SENT + "," + MANUALLY_STOPPED);
        DelayAlerterComponent.doMain(new String[]{"-c", pathToProperties});
        MimeMessage[] receivedMessages = greenMail.getReceivedMessages();
        //There could be other batches that trigger emails so check that there isn't one from us
        boolean found = false;
        for (MimeMessage message: receivedMessages) {
            if (GreenMailUtil.getBody(message).contains(batchId)) {
                found = true;
            }
        }
        assertFalse(found);
        assertFalse(batchContainsEvent(batchId, roundTrip, DelayAlerterComponent.EMAIL_SENT_EVENT));
    }

    /**
     * Waits for a given event to be either present or absent in SBOI.
     * @param batchId
     * @param roundTrip
     * @param eventId
     * @param isPresent
     */
    private void waitForEvent(String batchId, Integer roundTrip, String eventId, boolean isPresent) throws InterruptedException {
        Batch localBatch = new Batch(batchId, roundTrip);
        String type;
        if (isPresent) {
            type = "created";
        } else {
            type = "deleted";
        }
        logger.debug("Waiting for event {} {} to be {}.", localBatch.getFullID(), eventId, type);
        nsleeps = 0;

        while (true) {
            boolean eventPresent = false;
            Iterator<Batch> batchIterator = null;
            try {
                EventTrigger.Query<Batch> query = new EventTrigger.Query();
                query.getPastSuccessfulEvents().add(eventId);
                query.getItems().add(localBatch);
                batchIterator = sboi.getTriggeredItems(query);
                while (batchIterator.hasNext()) {
                    final Batch batch = batchIterator.next();
                    if (batch.getBatchID().equals(batchId) && batch.getRoundTripNumber().equals(roundTrip)) {
                        eventPresent = true;
                    }
                }
                boolean conditionSatisfied = (eventPresent && isPresent) || (!eventPresent && !isPresent);
                if (conditionSatisfied) {
                    logger.debug("Event {} {} {}.", localBatch.getFullID(), eventId,  type);
                    return;
                }
            } catch (CommunicationException e) {
                logger.debug(e.getMessage()); //Expected during reindexing.
            }
            Thread.sleep(sleep);
            nsleeps++;
            logger.debug("nsleeps = {}/{}.", nsleeps, maxSleeps);
            if (nsleeps > maxSleeps) throw new RuntimeException("SBOI not updated after " + maxSleeps*sleep/1000L + " seconds");
        }

    }

    private boolean batchContainsEvent(String batchId, Integer roundTrip, String eventId) throws Exception {
        Batch batch = null;
        try {
            batch = domsEventClient.getItemFromFullID(Batch.formatFullID(batchId,roundTrip));
        } catch (CommunicationException e) {
            return false;
        }
        for(Event event: batch.getEventList()) {
            if (event.getEventID().equals(eventId)) {
                return true;
            }
        }
        return false;
    }

}







src/main/java/dk/statsbiblioteket/newspaper/delayalerter/SimpleMailer.java
package dk.statsbiblioteket.newspaper.delayalerter;

import org.apache.zookeeper.server.SessionTracker;

import javax.mail.Message;
import javax.mail.MessagingException;
import javax.mail.Session;
import javax.mail.Transport;
import javax.mail.internet.AddressException;
import javax.mail.internet.InternetAddress;
import javax.mail.internet.MimeMessage;
import java.util.List;
import java.util.Properties;

/**
 * A simple mail-sending utility.
 */
public class SimpleMailer {

    private String from;
    private String host;
    private String port;

    /**
     * Constructor for this class.
     * @param from the from address field for emails.
     * @param host the smtp host to use.
     * @param port the smtp port of the host.
     */
    public SimpleMailer(String from, String host, String port) {
        this.from = from;
        this.host = host;
        this.port = port;
    }

    /**
     * Send a mail.
     * @param to A list of recipients.
     * @param subject The text of the email subject.
     * @param text The text of the email.
     * @throws MessagingException
     */
    public void sendMail(List<String> to, String subject, String text) throws MessagingException {
        Properties props = new Properties();
        props.setProperty("mail.smtp.host", host);
        props.setProperty("mail.smtp.port", port);
        javax.mail.Session session = Session.getDefaultInstance(props);
        MimeMessage message = new MimeMessage(session);
        message.setFrom(new InternetAddress(from));
        for (String recipient: to) {
            message.addRecipient(Message.RecipientType.TO, new InternetAddress(recipient));
        }
        message.setSubject(subject);
        message.setText(text);
        Transport.send(message);
    }
}


src/main/java/dk/statsbiblioteket/newspaper/delayalerter/DelayAlertMailer.java
package dk.statsbiblioteket.newspaper.delayalerter;

import org.apache.zookeeper.server.SessionTracker;

import javax.mail.Message;
import javax.mail.MessagingException;
import javax.mail.Session;
import javax.mail.Transport;
import javax.mail.internet.AddressException;
import javax.mail.internet.InternetAddress;
import javax.mail.internet.MimeMessage;
import java.util.List;
import java.util.Properties;

/**
 * A simple mail-sending utility.
 */
public class DelayAlertMailer {

    private String from;
    private String host;
    private String port;

    /**
     * Constructor for this class.
     * @param from the from address field for emails.
     * @param host the smtp host to use.
     * @param port the smtp port of the host.
     */
    public DelayAlertMailer(String from, String host, String port) {
        this.from = from;
        this.host = host;
        this.port = port;
    }

    /**
     * Send a mail.
     * @param to A list of recipients.
     * @param subject The text of the email subject.
     * @param text The text of the email.
     * @throws MessagingException
     */
    public void sendMail(List<String> to, String subject, String text) throws MessagingException {
        Properties props = new Properties();
        props.setProperty("mail.smtp.host", host);
        props.setProperty("mail.smtp.port", port);
        javax.mail.Session session = Session.getDefaultInstance(props);
        MimeMessage message = new MimeMessage(session);
        message.setFrom(new InternetAddress(from));
        for (String recipient: to) {
            message.addRecipient(Message.RecipientType.TO, new InternetAddress(recipient));
        }
        message.setSubject(subject);
        message.setText(text);
        Transport.send(message);
    }
}


src/main/java/dk/statsbiblioteket/newspaper/delayalerter/DelayAlerterConfigConstants.java
package dk.statsbiblioteket.newspaper.delayalerter;

/**
 * Class containing constants specifying the Property keys for this component.
 */
public final class DelayAlerterConfigConstants {

    /**
     * The number of days from when a batch is received until it is expected to be approved.
     */
    public static final String DELAY_ALERT_DAYS = "delay.alert.days";

    /**
     * A comma-separated list of email addresses to which alert messages will be sent.
     */
    public static final String DELAY_ALERT_EMAIL_ADDRESSES = "delay.alert.email.addresses";

    /**
     * The smtp host.
     */
    public static final String SMTP_HOST = "delay.alert.smtp.host";

    /**
     * The smtp port.
     */
    public static final String SMTP_PORT = "delay.alert.smtp.port";

    /**
     * The "from" email address.
     */
    public static final String EMAIL_FROM_ADDRESS = "delay.alert.email.from.address";

    /**
     * Private constructor as this class should not be instantiated.
     */
    private DelayAlerterConfigConstants() {
    }
}


src/main/java/dk/statsbiblioteket/newspaper/delayalerter/DelayAlerterComponent.java
package dk.statsbiblioteket.newspaper.delayalerter;

import dk.statsbiblioteket.doms.central.connectors.fedora.pidGenerator.PIDGeneratorException;
import dk.statsbiblioteket.medieplatform.autonomous.AbstractRunnableComponent;
import dk.statsbiblioteket.medieplatform.autonomous.AutonomousComponentUtils;
import dk.statsbiblioteket.medieplatform.autonomous.Batch;
import dk.statsbiblioteket.medieplatform.autonomous.CallResult;
import dk.statsbiblioteket.medieplatform.autonomous.Event;
import dk.statsbiblioteket.medieplatform.autonomous.ResultCollector;
import dk.statsbiblioteket.medieplatform.autonomous.RunnableComponent;
import dk.statsbiblioteket.medieplatform.autonomous.NewspaperBatchAutonomousComponentUtils;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.mail.MessagingException;
import javax.xml.bind.JAXBException;

import java.io.IOException;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.Properties;

/**
 * This is an autonomous component which checks if a batch has been active for too long without reaching
 * a completed status. Specifically it looks for batches with a "Data_Received" event but without an
 * "Approved" event and for which the "Data_Received" is more than XX days old. If this is the case
 * then it sends a warning email to a list of addresses.
 *
 * The component will only create an event to indicate that it is complete if it actually (successfully) sends a
 * warning email. In this way it can be rerun multiple times for a given batch-roundtrip.
 */
public class DelayAlerterComponent extends AbstractRunnableComponent<Batch> {

    private SimpleMailer mailer;

    public static String EMAIL_SENT_EVENT = "Warning_Email_Sent";

    private static Logger log = LoggerFactory.getLogger(DelayAlerterComponent.class);

    /**
     * This is the main method for this autonomous component. It's arguments should be
     * -c \<config.properties>
     *
     * @param args the arguments.
     * @throws IOException
     */
    public static void main(String[] args) throws IOException, JAXBException, PIDGeneratorException {
        System.exit(doMain(args));
    }

    static int doMain(String[] args) throws IOException {
        log.info("Starting with args {}", new Object[]{args});
        Properties properties = AutonomousComponentUtils.parseArgs(args);
        SimpleMailer mailer = new SimpleMailer(
                properties.getProperty(DelayAlerterConfigConstants.EMAIL_FROM_ADDRESS),
                properties.getProperty(DelayAlerterConfigConstants.SMTP_HOST),
                properties.getProperty(DelayAlerterConfigConstants.SMTP_PORT));
        RunnableComponent component = new DelayAlerterComponent(properties, mailer);
        CallResult result = NewspaperBatchAutonomousComponentUtils.startAutonomousComponent(properties, component);
        log.info(result.toString());
        return result.containsFailures();
    }

    /**
     * Constructor for the class.
     * @param properties Properties object containing all the necessary keys for this component.
     * @param mailer  A mailer to be used in sending an alert-message if necessary.
     */
    public DelayAlerterComponent(Properties properties, SimpleMailer mailer ) {
        super(properties);
        this.mailer = mailer;
    }

    @Override
    public String getEventID() {
        return EMAIL_SENT_EVENT;
    }

    @Override
    public void doWorkOnItem(Batch batch, ResultCollector resultCollector) throws Exception {
        List<Event> events=  batch.getEventList();
        for (Event event: events) {
            if (event.getEventID().equals("Data_Received")) {
                processDataReceivedEvent(batch, resultCollector, event);
            }
        }
    }

    /**
     * This method checks if the processing has taken too long. If it has, the event is passed on to sendAlertMail() for
     * further processing. Otherwise the resultCollector is set to non-preservable and the method just returns,
     * @param batch
     * @param resultCollector
     * @param event
     * @throws MessagingException
     */
    private void processDataReceivedEvent(Batch batch, ResultCollector resultCollector, Event event) throws MessagingException {
        Date now = new Date();
        Date receivedDate = event.getDate();
        final int delay = Integer.parseInt(
                getProperties().getProperty(DelayAlerterConfigConstants.DELAY_ALERT_DAYS));
        Long alertPeriod = delay *24*3600*1000L;
        log.debug("\nBatch:" + batch.getFullID() +"\nDate received: " + receivedDate + "\nNow:" + now + "\nDelay alert (days) " + delay);
        if (now.getTime() - receivedDate.getTime() > alertPeriod) {
            try {
                sendAlertMail(batch, resultCollector, event);
                resultCollector.setPreservable(true);
            } catch (MessagingException e) {
                log.error("Failed to send mail.", e);
                resultCollector.setPreservable(false);
                throw(e);
            }
        } else {
            log.debug("Not sending mail.");
            resultCollector.setPreservable(false);
            return;
        }
    }

    /**
     * Send the email alert.
     * @param batch
     * @param resultCollector
     * @param event
     * @throws MessagingException
     */
    private void sendAlertMail(Batch batch, ResultCollector resultCollector, Event event) throws MessagingException {
        String subject = "[Newspaper Delay Alert]" + batch.getFullID();
        String text = "Batch roundtrip " + batch.getFullID() + " was received at " + event.getDate() +
                "\n but has not yet been approved or rejected.";
        String[] mailRecipients = getProperties().getProperty(DelayAlerterConfigConstants.DELAY_ALERT_EMAIL_ADDRESSES).split(",");
        log.debug("Sending mail '" + text + "' to " + Arrays.asList(mailRecipients));
        mailer.sendMail(Arrays.asList(mailRecipients), subject, text);
    }
}
